crates/scheduler-server/src/handlers/metrics.rs

```diff
@@ -0,0 +1,158 @@
+use tokio::time::Instant;
+use tracing::{
+    span::{Attributes, Id},
+    Subscriber,
+};
+use tracing_subscriber::{layer::Context, registry::LookupSpan, Layer};
+
+lazy_static! {
+    static ref BUSY_HISTOGRAMS: EnumMap<FutureTimingMetric, Histogram> = enum_map! {
+        FutureTimingMetric::BuildingVsocksConnection => histogram!(format!("{}_busy", <&str>::from(FutureTimingMetric::BuildingVsocksConnection))),
+        FutureTimingMetric::SendingTeeRequest => histogram!(format!("{}_busy", <&str>::from(FutureTimingMetric::SendingTeeRequest))),
+        FutureTimingMetric::Health => histogram!(format!("{}_busy", <&str>::from(FutureTimingMetric::Health))),
+        FutureTimingMetric::GenerateProof => histogram!(format!("{}_busy", <&str>::from(FutureTimingMetric::GenerateProof))),
+        FutureTimingMetric::TeePublicKey => histogram!(format!("{}_busy", <&str>::from(FutureTimingMetric::TeePublicKey))),
+    };
+    static ref IDLE_HISTOGRAMS: EnumMap<FutureTimingMetric, Histogram> = enum_map! {
+        FutureTimingMetric::BuildingVsocksConnection => histogram!(format!("{}_idle", <&str>::from(FutureTimingMetric::BuildingVsocksConnection))),
+        FutureTimingMetric::SendingTeeRequest => histogram!(format!("{}_idle", <&str>::from(FutureTimingMetric::SendingTeeRequest))),
+        FutureTimingMetric::Health => histogram!(format!("{}_idle", <&str>::from(FutureTimingMetric::Health))),
+        FutureTimingMetric::GenerateProof => histogram!(format!("{}_idle", <&str>::from(FutureTimingMetric::GenerateProof))),
+        FutureTimingMetric::TeePublicKey => histogram!(format!("{}_idle", <&str>::from(FutureTimingMetric::TeePublicKey))),
+    };
+}
+
+#[derive(Debug, Clone, Copy, EnumIter, EnumString, IntoStaticStr, Enum)]
+pub enum FutureTimingMetric {
+    #[strum(serialize = "Building_VSOCK_connection")]
+    BuildingVsocksConnection,
+    #[strum(serialize = "Sending_TEE_request")]
+    SendingTeeRequest,
+    #[strum(serialize = "health")]
+    Health,
+    #[strum(serialize = "generate_proof")]
+    GenerateProof,
+    #[strum(serialize = "tee_public_key")]
+    TeePublicKey,
+}
+
+impl FutureTimingMetric {
+    pub fn by_name(name: &str) -> Option<Self> {
+        name.parse().ok()
+    }
+
+    pub const fn name(&self) -> &'static str {
+        match self {
+            FutureTimingMetric::BuildingVsocksConnection => "Building_VSOCK_connection",
+            FutureTimingMetric::SendingTeeRequest => "Sending_TEE_request",
+            FutureTimingMetric::Health => "health",
+            FutureTimingMetric::GenerateProof => "generate_proof",
+            FutureTimingMetric::TeePublicKey => "tee_public_key",
+        }
+    }
+
+    pub fn busy_histogram(&self) -> &'static Histogram {
+        &BUSY_HISTOGRAMS[*self]
+    }
+
+    pub fn idle_histogram_name(&self) -> &'static Histogram {
+        &IDLE_HISTOGRAMS[*self]
+    }
+}
+
+/// A tracing_subscriber layer that collects timing metrics for spans.
```

Comment:
Ideally code from this line till end of the file,  should be refactored to some common place, for starters `scheduler-common`, so they can be then re-used by prover server crate.